File: tools/physical/physical_dynamic_preflight.js
'use strict';

/**
 * Conservative no-effect safety proof for dynamic physical Runtime v2 programs.
 *
 * This module deliberately does not execute the student program and exposes no
 * backend, sensor, command or authority surface. It first delegates language
 * validity to the existing shared interpreter, then proves the physical safety
 * envelope for every syntactically reachable control-flow path. Conditions are
 * treated as nondeterministic: both branches are analysed, and repeat bodies are
 * analysed for every statically bounded iteration. The resulting altitude
 * interval is therefore an over-approximation; uncertainty rejects rather than
 * defers an unsafe path until after takeoff.
 */

const path = require('path');
const ROOT = path.resolve(__dirname, '../..');
const Interpreter = require(path.join(
  ROOT,
  'plugins/robot_windows/blockly/webeeblocks/interpreter.js'
));

const MIN_ALTITUDE_M = 0.2;
const MAX_ALTITUDE_M = 1.5;
const MIN_MOVE_DISTANCE_M = 0.10;
const MAX_MOVE_DISTANCE_M = 2.00;
const MIN_VERTICAL_DISTANCE_M = 0.10;
const MAX_VERTICAL_DISTANCE_M = 0.80;
const MIN_TURN_DEG = 1.0;
const MAX_TURN_DEG = 179.0;
const MIN_WAIT_SECONDS = 0.1;
const MAX_WAIT_SECONDS = 5.0;
const MIN_HORIZONTAL_SPEED_M_S = 0.10;
const MAX_HORIZONTAL_SPEED_M_S = 0.35;
const LIGHT_COLORS = new Set(['off', 'red', 'green', 'blue', 'yellow', 'white']);

function fail(message) {
  throw new Error('physical dynamic preflight: ' + message);
}

function exactKeys(value, keys, label) {
  if (!value || typeof value !== 'object' || Array.isArray(value))
    fail(label + ' must be an object');
  const actual = Object.keys(value).sort();
  const expected = keys.slice().sort();
  if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index]))
    fail(label + ' contains unsupported fields');
}

function exactOneOfKeySets(value, keySets, label) {
  for (const keys of keySets) {
    const actual = Object.keys(value).sort();
    const expected = keys.slice().sort();
    if (actual.length === expected.length && actual.every((key, index) => key === expected[index]))
      return;
  }
  fail(label + ' contains unsupported fields');
}

function finiteNumber(value, label) {
  if (typeof value !== 'number' || !Number.isFinite(value))
    fail(label + ' must be a finite number');
  return value;
}

function bounded(value, label, minimum, maximum) {
  const number = finiteNumber(value, label);
  if (number < minimum || number > maximum)
    fail(label + ' violates established physical bounds');
  return number;
}

function exactVariable(variable) {
  exactKeys(variable, ['id', 'name'], 'variable reference');
  if (typeof variable.id !== 'string' || !variable.id)
    fail('variable reference id is invalid');
  if (typeof variable.name !== 'string' || !variable.name.trim())
    fail('variable reference name is invalid');
}

function exactExpression(expression, depth) {
  if (depth > 20)
    fail('expression nesting is too deep');
  if (!expression || typeof expression !== 'object' || Array.isArray(expression))
    fail('expression must be an object');
  switch (expression.kind) {
    case 'number':
      exactKeys(expression, ['kind', 'value'], 'number expression');
      finiteNumber(expression.value, 'number expression value');
      return;
    case 'range':
      exactKeys(expression, ['kind', 'direction', 'unit'], 'range expression');
      return;
    case 'variable_get':
      exactKeys(expression, ['kind', 'variable'], 'variable expression');
      exactVariable(expression.variable);
      return;
    case 'arithmetic':
    case 'compare':
    case 'logic':
      exactKeys(expression, ['kind', 'op', 'left', 'right'], expression.kind + ' expression');
      exactExpression(expression.left, depth + 1);
      exactExpression(expression.right, depth + 1);
      return;
    default:
      fail('unsupported expression kind ' + String(expression.kind));
  }
}

function requireAltitude(interval, context) {
  if (!interval || typeof interval.min !== 'number' || typeof interval.max !== 'number')
    fail('internal altitude interval is invalid');
  if (
    !Number.isFinite(interval.min) ||
    !Number.isFinite(interval.max) ||
    interval.min < MIN_ALTITUDE_M ||
    interval.max > MAX_ALTITUDE_M
  )
    fail(context + ' can leave the established 0.2-1.5 m nominal-altitude envelope');
  return interval;
}

function exactStatementShape(statement, nested) {
  if (!statement || typeof statement !== 'object' || Array.isArray(statement))
    fail('statement must be an object');
  const kind = statement.kind;
  if (nested && (kind === 'takeoff' || kind === 'land'))
    fail('takeoff and land are only allowed at top-level boundaries');
  switch (kind) {
    case 'takeoff':
      exactKeys(statement, ['kind', 'height_m'], 'takeoff statement');
      bounded(statement.height_m, 'takeoff height_m', MIN_ALTITUDE_M, MAX_ALTITUDE_M);
      return;
    case 'land':
      exactKeys(statement, ['kind'], 'land statement');
      return;
    case 'move':
      exactKeys(statement, ['kind', 'direction', 'distance_m'], 'move statement');
      bounded(statement.distance_m, 'move distance_m', MIN_MOVE_DISTANCE_M, MAX_MOVE_DISTANCE_M);
      return;
    case 'vertical':
      exactKeys(statement, ['kind', 'direction', 'distance_m'], 'vertical statement');
      bounded(statement.distance_m, 'vertical distance_m', MIN_VERTICAL_DISTANCE_M, MAX_VERTICAL_DISTANCE_M);
      return;
    case 'turn': {
      exactKeys(statement, ['kind', 'angle_deg'], 'turn statement');
      const angle = finiteNumber(statement.angle_deg, 'turn angle_deg');
      const magnitude = Math.abs(angle);
      if (magnitude < MIN_TURN_DEG || magnitude > MAX_TURN_DEG)
        fail('turn angle_deg violates established physical bounds');
      return;
    }
    case 'wait':
      exactKeys(statement, ['kind', 'seconds'], 'wait statement');
      bounded(statement.seconds, 'wait seconds', MIN_WAIT_SECONDS, MAX_WAIT_SECONDS);
      return;
    case 'set_speed':
      exactKeys(statement, ['kind', 'speed_m_s'], 'set_speed statement');
      bounded(
        statement.speed_m_s,
        'set_speed speed_m_s',
        MIN_HORIZONTAL_SPEED_M_S,
        MAX_HORIZONTAL_SPEED_M_S
      );
      return;
    case 'set_light':
      exactKeys(statement, ['kind', 'color'], 'set_light statement');
      if (typeof statement.color !== 'string' || !LIGHT_COLORS.has(statement.color))
        fail('set_light color violates established physical palette');
      return;
    case 'set_variable':
      exactKeys(statement, ['kind', 'variable', 'value'], 'set_variable statement');
      exactVariable(statement.variable);
      exactExpression(statement.value, 0);
      return;
    case 'if':
      exactOneOfKeySets(
        statement,
        [
          ['kind', 'condition', 'then'],
          ['kind', 'condition', 'then', 'else']
        ],
        'if statement'
      );
      exactExpression(statement.condition, 0);
      if (!Array.isArray(statement.then) || (statement.else !== undefined && !Array.isArray(statement.else)))
        fail('if branches must be statement arrays');
      return;
    case 'repeat':
      exactKeys(statement, ['kind', 'count', 'body'], 'repeat statement');
      if (!Number.isInteger(statement.count) || statement.count < 1 || statement.count > 20)
        fail('repeat count violates shared interpreter bounds');
      if (!Array.isArray(statement.body))
        fail('repeat body must be a statement array');
      return;
    default:
      fail('unsupported physical statement kind ' + String(kind));
  }
}

function analyseSequence(sequence, interval, depth) {
  if (!Array.isArray(sequence))
    fail('statement sequence must be an array');
  if (depth > 20)
    fail('statement nesting is too deep');
  let current = {min: interval.min, max: interval.max};
  for (const statement of sequence) {
    exactStatementShape(statement, true);
    switch (statement.kind) {
      case 'vertical': {
        const distance = statement.distance_m;
        const delta = statement.direction === 'up' ? distance : -distance;
        current = requireAltitude(
          {min: current.min + delta, max: current.max + delta},
          'vertical path'
        );
        break;
      }
      case 'if': {
        const thenResult = analyseSequence(statement.then, current, depth + 1);
        const elseResult = analyseSequence(statement.else || [], current, depth + 1);
        current = requireAltitude(
          {
            min: Math.min(thenResult.min, elseResult.min),
            max: Math.max(thenResult.max, elseResult.max)
          },
          'conditional path'
        );
        break;
      }
      case 'repeat':
        for (let iteration = 0; iteration < statement.count; iteration += 1)
          current = analyseSequence(statement.body, current, depth + 1);
        break;
      case 'takeoff':
      case 'land':
        fail('takeoff and land cannot occur inside the in-flight dynamic body');
        break;
      default:
        // Non-vertical effects, observations, variables and host-local state are
        // nominal-altitude neutral. Their exact physical bounds were checked above.
        break;
    }
  }
  return current;
}

function validatePhysicalProgram(ast) {
  // Language/control-flow validity remains owned by the shared interpreter.
  Interpreter.validateProgram(ast);

  exactKeys(ast, ['version', 'semantics', 'program'], 'physical AST envelope');
  if (!Array.isArray(ast.program) || ast.program.length < 2)
    fail('physical program must contain flight boundaries');

  const first = ast.program[0];
  const last = ast.program[ast.program.length - 1];
  exactStatementShape(first, false);
  exactStatementShape(last, false);
  if (first.kind !== 'takeoff')
    fail('physical program must begin with exact takeoff');
  if (last.kind !== 'land')
    fail('physical program must end with exact land');

  const initialAltitude = bounded(
    first.height_m,
    'takeoff height_m',
    MIN_ALTITUDE_M,
    MAX_ALTITUDE_M
  );
  const terminal = analyseSequence(
    ast.program.slice(1, -1),
    {min: initialAltitude, max: initialAltitude},
    0
  );

  return Object.freeze({
    initialAltitudeM: initialAltitude,
    terminalAltitudeMinM: terminal.min,
    terminalAltitudeMaxM: terminal.max,
    executionAuthority: false
  });
}

module.exports = Object.freeze({
  validatePhysicalProgram,
  MIN_ALTITUDE_M,
  MAX_ALTITUDE_M
});

File: tools/physical/physical_dynamic_preflight.py
#!/usr/bin/env python3
"""Conservative pre-takeoff safety envelope for dynamic physical programs.

This module is deliberately non-authority and emits no physical effect. It
consumes only the exact canonical AST binding already established by trusted
current-program provenance and proves that every dynamically reachable physical
path remains inside the already integrated action and nominal-altitude bounds.

It does not evaluate expressions or choose branches. `if` is treated as either
branch being reachable and `repeat` is expanded for its exact bounded count.
The shared Runtime interpreter remains the sole language evaluator; this module
only supplies a conservative physical safety proof before reset/takeoff.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import isfinite

import high_level_semantics
import high_level_timing
import physical_program_sequence
import takeoff_command

MAX_CONTROL_DEPTH = 20


class DynamicPhysicalPreflightError(RuntimeError):
    """Fail-closed error for an unsafe or unsupported dynamic physical program."""


@dataclass(frozen=True, slots=True)
class ReachablePhysicalEnvelope:
    """Bounded world-Z envelope proven across every admitted control-flow path."""

    ast_binding: str
    initial_altitude_m: float
    min_altitude_m: float
    max_altitude_m: float
    terminal_min_altitude_m: float
    terminal_max_altitude_m: float


@dataclass(frozen=True, slots=True)
class _Bounds:
    low: float
    high: float
    seen_low: float
    seen_high: float


def _error(message: str) -> DynamicPhysicalPreflightError:
    return DynamicPhysicalPreflightError(message)


def _exact_fields(statement: dict[str, object], fields: set[str], context: str) -> None:
    if set(statement) != fields:
        raise _error(context + " contains unsupported fields")


def _number(value: object, context: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise _error(context + " must be finite")
    parsed = float(value)
    if not isfinite(parsed):
        raise _error(context + " must be finite")
    return parsed


def _require_altitude(low: float, high: float) -> None:
    if (
        not isfinite(low)
        or not isfinite(high)
        or low < physical_program_sequence.MIN_NOMINAL_ALTITUDE_M
        or high > physical_program_sequence.MAX_NOMINAL_ALTITUDE_M
    ):
        raise _error(
            "a reachable physical path violates established 0.2-1.5 m nominal-altitude bounds"
        )


def _validate_action(statement: dict[str, object], bounds: _Bounds) -> _Bounds | None:
    kind = statement.get("kind")
    try:
        if kind == "move":
            _exact_fields(statement, {"kind", "direction", "distance_m"}, "move statement")
            direction = statement["direction"]
            if not isinstance(direction, str):
                raise _error("move direction is malformed")
            high_level_semantics.body_relative_move(
                direction,
                statement["distance_m"],
                0.0,
            )
            return bounds

        if kind == "vertical":
            _exact_fields(
                statement,
                {"kind", "direction", "distance_m"},
                "vertical statement",
            )
            direction = statement["direction"]
            if not isinstance(direction, str):
                raise _error("vertical direction is malformed")
            target = high_level_semantics.vertical_move(
                direction,
                statement["distance_m"],
            )
            low = bounds.low + target.z_m
            high = bounds.high + target.z_m
            _require_altitude(low, high)
            return _Bounds(
                low=low,
                high=high,
                seen_low=min(bounds.seen_low, low),
                seen_high=max(bounds.seen_high, high),
            )

        if kind == "turn":
            _exact_fields(statement, {"kind", "angle_deg"}, "turn statement")
            high_level_semantics.relative_turn(statement["angle_deg"])
            return bounds

        if kind == "wait":
            _exact_fields(statement, {"kind", "seconds"}, "wait statement")
            seconds = _number(statement["seconds"], "wait seconds")
            if (
                seconds < physical_program_sequence.MIN_WAIT_SECONDS
                or seconds > physical_program_sequence.MAX_WAIT_SECONDS
            ):
                raise _error("wait seconds violate established Runtime v2 bounds")
            return bounds

        if kind == "set_speed":
            _exact_fields(statement, {"kind", "speed_m_s"}, "set_speed statement")
            speed = _number(statement["speed_m_s"], "set_speed speed_m_s")
            if (
                speed < high_level_timing.MIN_HORIZONTAL_SPEED_M_S
                or speed > high_level_timing.MAX_HORIZONTAL_SPEED_M_S
            ):
                raise _error(
                    "set_speed violates established physical horizontal-speed bounds"
                )
            return bounds

        if kind == "set_light":
            _exact_fields(statement, {"kind", "color"}, "set_light statement")
            color = statement["color"]
            if (
                not isinstance(color, str)
                or color not in physical_program_sequence.LIGHT_COLORS
            ):
                raise _error("set_light color violates established Runtime v2 palette")
            return bounds
    except DynamicPhysicalPreflightError:
        raise
    except (
        TypeError,
        ValueError,
        high_level_semantics.HighLevelSemanticError,
    ) as exc:
        raise _error("reachable physical action violates integrated semantics") from exc

    return None


def _validate_sequence(
    sequence: object,
    bounds: _Bounds,
    *,
    depth: int,
) -> _Bounds:
    if depth > MAX_CONTROL_DEPTH:
        raise _error("dynamic physical control-flow nesting is too deep")
    if not isinstance(sequence, list):
        raise _error("dynamic physical statement sequence must be an array")

    current = bounds
    for statement in sequence:
        if not isinstance(statement, dict) or not isinstance(statement.get("kind"), str):
            raise _error("dynamic physical statement is malformed")

        action = _validate_action(statement, current)
        if action is not None:
            current = action
            continue

        kind = statement["kind"]
        if kind == "set_variable":
            _exact_fields(
                statement,
                {"kind", "variable", "value"},
                "set_variable statement",
            )
            variable = statement["variable"]
            if (
                not isinstance(variable, dict)
                or set(variable) != {"id", "name"}
                or not isinstance(variable.get("id"), str)
                or not variable["id"]
                or not isinstance(variable.get("name"), str)
                or not variable["name"].strip()
            ):
                raise _error("set_variable reference is malformed")
            if not isinstance(statement["value"], dict):
                raise _error("set_variable expression is malformed")
            continue

        if kind == "if":
            allowed = {"kind", "condition", "then", "else"}
            if set(statement) not in ({"kind", "condition", "then"}, allowed):
                raise _error("if statement contains unsupported fields")
            if not isinstance(statement["condition"], dict):
                raise _error("if condition is malformed")
            then_bounds = _validate_sequence(
                statement["then"],
                current,
                depth=depth + 1,
            )
            else_bounds = _validate_sequence(
                statement.get("else", []),
                current,
                depth=depth + 1,
            )
            current = _Bounds(
                low=min(then_bounds.low, else_bounds.low),
                high=max(then_bounds.high, else_bounds.high),
                seen_low=min(
                    current.seen_low,
                    then_bounds.seen_low,
                    else_bounds.seen_low,
                ),
                seen_high=max(
                    current.seen_high,
                    then_bounds.seen_high,
                    else_bounds.seen_high,
                ),
            )
            _require_altitude(current.low, current.high)
            continue

        if kind == "repeat":
            _exact_fields(statement, {"kind", "count", "body"}, "repeat statement")
            count_value = statement["count"]
            if (
                isinstance(count_value, bool)
                or not isinstance(count_value, (int, float))
                or not float(count_value).is_integer()
            ):
                raise _error("repeat count must be an integer")
            count = int(count_value)
            if count < 1 or count > 20:
                raise _error("repeat count violates established Runtime v2 bounds")
            if not isinstance(statement["body"], list):
                raise _error("repeat body is malformed")
            for _ in range(count):
                current = _validate_sequence(
                    statement["body"],
                    current,
                    depth=depth + 1,
                )
            continue

        if kind in {"takeoff", "land"}:
            raise _error("takeoff and land are only allowed at top-level boundaries")

        raise _error("unsupported dynamic physical statement kind: " + kind)

    return current


def validate_bound_dynamic_program(ast_binding: object) -> ReachablePhysicalEnvelope:
    """Prove every reachable physical path before reset/takeoff.

    Expression evaluation and branch selection remain exclusively owned by the
    shared Runtime interpreter. This proof assumes either branch of every `if`
    can be selected and therefore cannot be weakened by a sensor value.
    """

    try:
        takeoff = takeoff_command.derive_bound_takeoff_command(ast_binding)
        parsed = takeoff_command._parse_ast_binding(ast_binding)
    except takeoff_command.TakeoffCommandError as exc:
        raise _error(str(exc)) from exc

    program = parsed["program"]
    if not isinstance(program, list):
        raise _error("physical AST program is unavailable")
    final = program[-1]
    if (
        not isinstance(final, dict)
        or set(final) != {"kind"}
        or final.get("kind") != "land"
    ):
        raise _error("physical AST must end with the exact landing statement")

    initial = float(takeoff.height_m)
    _require_altitude(initial, initial)
    start = _Bounds(
        low=initial,
        high=initial,
        seen_low=initial,
        seen_high=initial,
    )
    terminal = _validate_sequence(
        program[1:-1],
        start,
        depth=0,
    )
    return ReachablePhysicalEnvelope(
        ast_binding=str(ast_binding),
        initial_altitude_m=initial,
        min_altitude_m=terminal.seen_low,
        max_altitude_m=terminal.seen_high,
        terminal_min_altitude_m=terminal.low,
        terminal_max_altitude_m=terminal.high,
    )
